### PR TITLE
Improve Rubygems declared license storage #1480

### PR DESCRIPTION
**Pull request by linexb**
See #1480 
Store the declared license as a list of strings and compute a license expression from these.

### CODE REVIEW

File: src/packagedcode/rubygems.py [hunk 1]
```diff
@@ -385,6 +398,22 @@ def build_rubygem_package(gem_data, download_url=None, package_url=None):
...
+    `licenses` values found in a package.
+    """
+    declared_licenses = []
+    if license:
+        declared_licenses.append(str(license).strip())
+    if licenses:
+        for lic in licenses:
+            if lic and lic.strip():
+                declared_licenses.append(lic.strip())
+    package.declared_license = declared_licenses
```
Comment (pombredanne) [inline src/packagedcode/rubygems.py]: Instead of updating the package, please return the list of declared licenses.

File: src/packagedcode/rubygems.py [hunk 2]
```diff
@@ -314,6 +320,13 @@ def build_rubygem_package(gem_data, download_url=None, package_url=None):
         declared_license=declared_license,
     )
 
+    # Since the gem spec doc is not clear https://guides.rubygems.org
+    # /specification-reference/#licenseo, we will treat a list of licenses and a
+    # conjunction for now (e.g. AND)
+    license = gem_data.get('license')
+    licenses = gem_data.get('licenses')
+    package = licenses_mapper(license, licenses, package)
```
Comment (pombredanne) [inline src/packagedcode/rubygems.py]: do not pass the package around.... instead get a declared_license list back and add it to the package creation above.... so this can and should be moved before the package creation

File: src/packagedcode/rubygems.py [hunk 3]
```diff
@@ -306,6 +311,7 @@ def build_rubygem_package(gem_data, download_url=None, package_url=None):
 
     declared_license = '\n'.join(declared_licenses) or None
 
+
```
Comment (pombredanne) [inline src/packagedcode/rubygems.py]: `declared_license = '\n'.join(declared_licenses) or None` 

.... this no longer makes sense. We want to explicitly keep a list of such licenses and no longer a string

File: src/packagedcode/utils.py [hunk 4]
```diff
@@ -146,3 +147,22 @@ def build_description(summary, description):
...
 
     return description
+
+def combine_expressions(expressions, licensing=Licensing()):
+    """
+    Return a combined license expression string with AND, given a list of
+    license expressions.
+
+    For example:
+    >>> a = 'mit'
```
Comment (pombredanne) [inline src/packagedcode/utils.py]: Can you add a test for what happens when expressions is None, an empty list or a string or a string with a single character?

File: src/packagedcode/utils.py [hunk 5]
```diff
@@ -146,3 +147,22 @@ def build_description(summary, description):
...
+    >>> b = 'gpl'
+    >>> combine_expressions([a, b])
+    'mit AND gpl'
+    """
+    if not expressions:
+        return
+    if len(expressions) == 1:
```
Comment (pombredanne) [inline src/packagedcode/utils.py]: what happens if expressions is `q` ? this tests alone does not make sense to me.

File: src/packagedcode/maven.py [hunk 6]
```diff
@@ -202,27 +203,16 @@ def compute_normalized_license(declared_license):
...
                 if detections:
-                    if len(detections) == 1:
-                        combined_expression = detections[0]
-                    else:
-                        expressions = [
-                            licensing.parse(le, simple=True) for le in detections]
-                        combined_expression = str(licensing.AND(*expressions))
-                    detected_licenses.append(combined_expression)
-
+                    combined_expression =  combine_expressions(detections)
```
[the remark this answers is not in the document]
Comment (pombredanne) [reply src/packagedcode/maven.py]: Actually my bad: there is no change t the logic here! :+1:

File: src/packagedcode/rubygems.py (hunk 2 above)
Comment (linexb) [reply src/packagedcode/rubygems.py]: Fixed.

File: src/packagedcode/rubygems.py (hunk 3 above)
Comment (linexb) [reply src/packagedcode/rubygems.py]: Fixed

File: src/packagedcode/rubygems.py [hunk 7]
```diff
@@ -167,6 +150,30 @@ def extra_root_dirs(cls):
...
+    """
+    if not declared_license:
+        return
+
+    detected_licenses = []
+
+    for declared in declared_license:
+        #TODO:  scancode convention is to put one license per line when there are
+        # multiple licenses as a list in the manifest.
+        # lines = [l for l in self.declared_license.splitlines(False) if l and l.strip()]
```
[the remark this answers is not in the document]
Comment (linexb) [reply src/packagedcode/rubygems.py]: Fixed.

File: src/packagedcode/rubygems.py (hunk 1 above)
Comment (linexb) [reply src/packagedcode/rubygems.py]: Fixed.

File: src/packagedcode/rubygems.py [hunk 8]
```diff
@@ -167,6 +150,30 @@ def extra_root_dirs(cls):
...
+        if declared:
```
[the remark this answers is not in the document]
Comment (linexb) [reply src/packagedcode/rubygems.py]: Agree, the licenses_mapper already checked the empty case, so it's impossible to be empty here. Fixed.

File: src/packagedcode/utils.py (hunk 4 above)
Comment (linexb) [reply src/packagedcode/utils.py]: Yep, I have added the tests on  the pypi branch, https://github.com/nexB/scancode-toolkit/commit/801bb5d92525fbc036e8fdab829ee41724f57f65

File: src/packagedcode/utils.py (hunk 5 above)
Comment (linexb) [reply src/packagedcode/utils.py]: @pombredanne Can you please explain more about q? Do you mean non list case? :)

Comment (pombredanne) [reply src/packagedcode/utils.py]: what if this expression is only a single character string and not a list?

**Comment by pombredanne** [comment]
Thanks. All merged
